cou/steps/hypervisor.py

```diff
@@ -84,25 +90,52 @@ def apps(self) -> list[OpenStackApplication]:
         return self._apps
 
     @property
-    def azs(self) -> list[HypervisorGroup]:
+    def azs(self) -> dict[str, HypervisorGroup]:
         """Returns a list of AZs defined in individual applications.
 
-        Each AZ contains a sorted list hypervisors, where each hypervisor has sorted list of units.
-        The order of units depends on the order in which they needs to be upgraded.
+        Each AZ contains a dictionary of application name and all units in the AZ.
         eg.
         az1:
-        - hypervisor0:
+        - cinder:
           - cinder/0
-          - nova-compute/0
-        - hypervisor1:
           - cinder/1
+          - cinder/2
+        - nova-compute
+          - nova-compute/0
           - nova-compute/1
+          - nova-compute/2
+        ...
+        az2
+        - cinder
+          -cinder/3
+        ...
         ...
 
-        :return: List of availability zones.
-        :rtype: list[HypervisorGroup]
+        :return: dictionary with key as az name and value as HypervisorGroup
+        :rtype: dict[str, HypervisorGroup]
+        :raises ApplicationError: if there is unit without az defined
         """
-        raise NotImplementedError
+        azs = {}
+        for app in self.apps:
+            for unit in app.units.values():
+                az = unit.machine.az
+
+                if az is None:
+                    raise ApplicationError(
+                        f"The application {app.name} has a machine {unit.machine} without az"
+                    )
+
+                if az not in azs:
+                    # define empty hypervisor group for az
+                    azs[az] = HypervisorGroup(name=az, apps={})
+
+                if app.name not in azs[az].apps:
+                    # define empty list of units for each app, so unit can be added
+                    azs[az].apps[app.name] = []
```

Comment:
Would be simpler if we do:
```suggestion
                if az not in azs:
                    # define empty hypervisor group for az
                    azs[az] = HypervisorGroup(name=az, apps={app.name: []})
                elif app.name not in azs[az].apps:
                    # define empty list of units for each app, so unit can be added
                    azs[az].apps[app.name] = []
```

Reply:
That will not work, since with first if you created whole ` azs[az]` and in second you add empty list for app. 

We can do it, but like this:
```python
if az not in azs:
  # define empty hypervisor group for az
  azs[az] = HypervisorGroup(name=az, apps={app.name: []})
  azs[az].apps[app.name] = []
elif app.name not in azs[az].apps:
  # define empty list of units for each app, so unit can be added
  azs[az].apps[app.name] = []
```
and there is no benefit doing it like that.

Reply:
I don't understand what you mean here: with my approach the empty list is already added in for app in `azs[az] = HypervisorGroup(name=az, apps={app.name: []})`. Let's discuss it tomorrow in the sync.

Reply:
Ah I see that you are using defaultdict instead, which looks better.